### PR TITLE
Cuda 10.1 support

### PR DESCRIPTION
**Pull request by nluehr**
Enables building TensorFlow with CUDA 10.1.
Addresses #26150

### CODE REVIEW

**Comment by yifeif** [comment]
@chsigg, do you mind taking a look at this? Thanks!

File: tensorflow/stream_executor/cuda/cuda_stub.cc [hunk 1]
```diff
@@ -96,4 +96,8 @@ typedef void(CUDA_CB* CUhostFn)(void* userData);
 // We only need one stub implementation. Calling a function that is not yet
 // available in the given CUDA version will return
 // CUDA_ERROR_SHARED_OBJECT_INIT_FAILED.
+#if CUDA_VERSION < 10010
```
Comment (chsigg) [inline tensorflow/stream_executor/cuda/cuda_stub.cc]: The comment doesn't match the implementation anymore.

**Review by chsigg — COMMENTED** [review]
Thanks for this PR Nathan.

Could you explain the cuBLAS changes that requires splitting CUDA_VERSION and CUDA_LIB_VERSION? Or should this be done independently of 10.1? If so, may be that should be done as a separate PR?

You mentioned some constness changes in the CUDA API that requires us to add extra wrappers. I'm randomly guessing that the changes were in the graph API, and I'm wondering if it would be easier for now to just remove them from the wrapper because we don't use the graph API in TensorFlow.

**Comment by nluehr** [comment]
@chsigg Starting with CUDA 10.1 cublas is installed into the system /usr/lib and /usr/include directories rather than being located with the toolkit. Also, a build number is being added to the cublas libs, so they won't necessarily match the toolkit version. See https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cublas-new-features

Reducing the coverage of the stub library should improve future maintainability as well. I'll update the PR shortly.

File: tensorflow/stream_executor/cuda/cuda_stub.cc (hunk 1 above)
Comment (nluehr) [reply tensorflow/stream_executor/cuda/cuda_stub.cc]: Fixed.

**Comment by nluehr** [comment]
@chsigg anything else needed on this?

File: third_party/gpus/cuda_configure.bzl [hunk 2]
```diff
@@ -1295,6 +1312,19 @@ def _create_local_cuda_repository(repository_ctx):
     else:
         copy_rules.append("filegroup(name = 'cudnn-include')\n")
 
+    # Copy cublas headers if not installed to CUDA_TOOLKIT_PATH.
+    if not any([file.endswith("cublas_v2.h") for file in included_files]):
+        cublas_header_dir = "/usr/include"
```
Comment (chsigg) [inline third_party/gpus/cuda_configure.bzl]: Would it be possible to check where these files can be found in configure.py, similar to how we do it for cuDNN and TensorRT? I'm suspecting the path will be different on Windows.

**Review by chsigg — COMMENTED** [review]
Would you mind not changing the filename for cuda_10_0.inc? Or if you prefer, maybe change it to cuda_10.inc.

**Comment by chsigg** [comment]
Hi Nathan. Thanks for updating the PR, and sorry for the delay.

**Comment by mmstick** [comment]
Would this still be compatible for those who are packaging CUDA 9.X + 10.X to place them in their own path? For Pop!_OS, I have each CUDA SDK to store their files in `/usr/lib/cuda-${version}`, with `update-alternatives` making it possible to switch active toolchains from a symlink at `/usr/lib/cuda`. This enables multiple toolchains to be installed at the same time.

We've packaged CUDA 10.1 in this manner as well:

```sh
# ls /usr/lib/cuda-10.1/lib64
libaccinj64.so		      libcufftw.so.10.1.105    libnppc_static.a        libnppig.so.10.1.105   libnvblas.so.10
libaccinj64.so.10.1	      libcufftw_static.a       libnppial.so	       libnppig_static.a      libnvblas.so.10.1.0.105
libaccinj64.so.10.1.105       libcuinj64.so	       libnppial.so.10	       libnppim.so	      libnvgraph.so
libcublasLt.so		      libcuinj64.so.10.1       libnppial.so.10.1.105   libnppim.so.10	      libnvgraph.so.10
libcublasLt.so.10	      libcuinj64.so.10.1.105   libnppial_static.a      libnppim.so.10.1.105   libnvgraph.so.10.1.105
libcublasLt.so.10.1.0.105     libculibos.a	       libnppicc.so	       libnppim_static.a      libnvgraph_static.a
libcublasLt_static.a	      libcurand.so	       libnppicc.so.10	       libnppist.so	      libnvjpeg.so
libcublas.so		      libcurand.so.10	       libnppicc.so.10.1.105   libnppist.so.10	      libnvjpeg.so.10
libcublas.so.10		      libcurand.so.10.1.105    libnppicc_static.a      libnppist.so.10.1.105  libnvjpeg.so.10.1.105
libcublas.so.10.1.0.105       libcurand_static.a       libnppicom.so	       libnppist_static.a     libnvjpeg_static.a
libcublas_static.a	      libcusolver.so	       libnppicom.so.10        libnppisu.so	      libnvrtc-builtins.so
libcudadevrt.a		      libcusolver.so.10        libnppicom.so.10.1.105  libnppisu.so.10	      libnvrtc-builtins.so.10.1
libcudart.so		      libcusolver.so.10.1.105  libnppicom_static.a     libnppisu.so.10.1.105  libnvrtc-builtins.so.10.1.105
libcudart.so.10.1	      libcusolver_static.a     libnppidei.so	       libnppisu_static.a     libnvrtc.so
libcudart.so.10.1.105	      libcusparse.so	       libnppidei.so.10        libnppitc.so	      libnvrtc.so.10.1
libcudart_static.a	      libcusparse.so.10        libnppidei.so.10.1.105  libnppitc.so.10	      libnvrtc.so.10.1.105
libcufft.so		      libcusparse.so.10.1.105  libnppidei_static.a     libnppitc.so.10.1.105  libnvToolsExt.so
libcufft.so.10		      libcusparse_static.a     libnppif.so	       libnppitc_static.a     libnvToolsExt.so.1
libcufft.so.10.1.105	      liblapack_static.a       libnppif.so.10	       libnpps.so	      libnvToolsExt.so.1.0.0
libcufft_static.a	      libmetis_static.a        libnppif.so.10.1.105    libnpps.so.10	      libOpenCL.so
libcufft_static_nocallback.a  libnppc.so	       libnppif_static.a       libnpps.so.10.1.105    libOpenCL.so.1
libcufftw.so		      libnppc.so.10	       libnppig.so	       libnpps_static.a       libOpenCL.so.1.1
libcufftw.so.10		      libnppc.so.10.1.105      libnppig.so.10	       libnvblas.so	      stubs
# ls /usr/lib/cuda-10.1/include
builtin_types.h		      curand_discrete2.h	     host_defines.h			       nvToolsExtCudaRt.h
channel_descriptor.h	      curand_discrete.h		     library_types.h			       nvToolsExt.h
CL			      curand_globals.h		     math_constants.h			       nvToolsExtMeta.h
common_functions.h	      curand.h			     math_functions.h			       nvToolsExtSync.h
cooperative_groups.h	      curand_kernel.h		     mma.h				       nvtx3
cooperative_groups_helpers.h  curand_lognormal.h	     nppcore.h				       sm_20_atomic_functions.h
crt			      curand_mrg32k3a.h		     nppdefs.h				       sm_20_atomic_functions.hpp
cublas_api.h		      curand_mtgp32dc_p_11213.h      npp.h				       sm_20_intrinsics.h
cublas.h		      curand_mtgp32.h		     nppi_arithmetic_and_logical_operations.h  sm_20_intrinsics.hpp
cublasLt.h		      curand_mtgp32_host.h	     nppi_color_conversion.h		       sm_30_intrinsics.h
cublas_v2.h		      curand_mtgp32_kernel.h	     nppi_compression_functions.h	       sm_30_intrinsics.hpp
cublas_version.h	      curand_normal.h		     nppi_computer_vision.h		       sm_32_atomic_functions.h
cublasXt.h		      curand_normal_static.h	     nppi_data_exchange_and_initialization.h   sm_32_atomic_functions.hpp
cuComplex.h		      curand_philox4x32_x.h	     nppi_filtering_functions.h		       sm_32_intrinsics.h
cuda_device_runtime_api.h     curand_poisson.h		     nppi_geometry_transforms.h		       sm_32_intrinsics.hpp
cudaEGL.h		      curand_precalc.h		     nppi.h				       sm_35_atomic_functions.h
cuda_egl_interop.h	      curand_uniform.h		     nppi_linear_transforms.h		       sm_35_intrinsics.h
cuda_fp16.h		      cusolver_common.h		     nppi_morphological_operations.h	       sm_60_atomic_functions.h
cuda_fp16.hpp		      cusolverDn.h		     nppi_statistics_functions.h	       sm_60_atomic_functions.hpp
cudaGL.h		      cusolverRf.h		     nppi_support_functions.h		       sm_61_intrinsics.h
cuda_gl_interop.h	      cusolverSp.h		     nppi_threshold_and_compare_operations.h   sm_61_intrinsics.hpp
cuda.h			      cusolverSp_LOWLEVEL_PREVIEW.h  npps_arithmetic_and_logical_operations.h  sobol_direction_vectors.h
cudalibxt.h		      cusparse.h		     npps_conversion_functions.h	       surface_functions.h
cuda_occupancy.h	      cusparse_v2.h		     npps_filtering_functions.h		       surface_functions.hpp
cuda_profiler_api.h	      device_atomic_functions.h      npps.h				       surface_indirect_functions.h
cudaProfiler.h		      device_atomic_functions.hpp    npps_initialization.h		       surface_indirect_functions.hpp
cudart_platform.h	      device_double_functions.h      npps_statistics_functions.h	       surface_types.h
cuda_runtime_api.h	      device_functions.h	     npps_support_functions.h		       texture_fetch_functions.h
cuda_runtime.h		      device_launch_parameters.h     nppversion.h			       texture_fetch_functions.hpp
cuda_surface_types.h	      device_types.h		     nvblas.h				       texture_indirect_functions.h
cuda_texture_types.h	      driver_functions.h	     nvfunctional			       texture_indirect_functions.hpp
cudaVDPAU.h		      driver_types.h		     nvgraph.h				       texture_types.h
cuda_vdpau_interop.h	      fatBinaryCtl.h		     nvjpeg.h				       thrust
cufft.h			      fatbinary.h		     nvml.h				       vector_functions.h
cufftw.h		      fatbinary_section.h	     nvrtc.h				       vector_functions.hpp
cufftXt.h		      host_config.h		     nvToolsExtCuda.h			       vector_types.h
```

**Comment by chsigg** [comment]
It's not clear to me whether you want to build or run TensorFlow with
different CUDA versions, but I think this PR should not affect either (for
building, assuming we allow specifying the cuBLAS include directory during
./configure)

**Comment by mmstick** [comment]
That would be both building and running TF with different CUDA versions. We compile a new TF package for each CUDA version + a CPU package, so that `tensorflow-1.13-cuda-10.1` would be a TF package built with `system76-cuda-10.1` & co. as dependencies. This also allows installing multiple versions of TF and toggling between the active version symlinked at `/usr/lib/tensorflow`.

File: third_party/gpus/cuda_configure.bzl [hunk 3]
```diff
@@ -528,9 +529,15 @@ def _cuda_version(repository_ctx, cuda_toolkit_path, cpu_value):
...
+        lib_version = version
+        if (int(version_parts[0]) == 10 and int(version_parts[1]) >= 1) or (int(version_parts[0]) > 10):
+            lib_version = "64_%s" % (version_parts[0])
     else:
         version = "%s.%s" % (version_parts[0], version_parts[1])
-    return version
+        lib_version = version
+        if (int(version_parts[0]) == 10 and int(version_parts[1]) >= 1) or (int(version_parts[0]) > 10):
+            lib_version = "%s" % (version_parts[0])
+    return (version, lib_version)
```
Comment (chsigg) [inline third_party/gpus/cuda_configure.bzl]: Please update the macro documentation.

**Comment by chsigg** [comment]
Hi Nathan. Sorry that this hasn't made any progress. I will try to split this up a little and get it submitted as multiple CLs.

**Comment by trevor-m** [comment]
FYI, the version for cusparse needs to be updated as well: https://github.com/tensorflow/tensorflow/blob/master/third_party/gpus/cuda_configure.bzl#L907

**Comment by edrozenberg** [comment]
> @chsigg Starting with CUDA 10.1 cublas is installed into the system /usr/lib and /usr/include directories rather than being located with the toolkit. Also, a build number is being added to the cublas libs, so they won't necessarily match the toolkit version. See https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html#cublas-new-features
> 
> Reducing the coverage of the stub library should improve future maintainability as well. I'll update the PR shortly.

On my systems I consolidate everything in /opt/nvidia/cuda(/lib64) because I don't like how the Nvidia installer spews a bunch of stuff into various system directories.

Hopefully whatever build changes are made to TF won't be making assumptions that things are located in the "standard" places the latest flavor of the Nvidia installer decides to use.

My hope is the approach below I use will continue to work.

```
$ cat /etc/ld.so.conf.d/cuda.conf 

/opt/nvidia/cuda/lib64
/opt/nvidia/cuda/extras/CUPTI/lib64
/opt/nvidia/nccl/lib

$ ldconfig
```

**Comment by edrozenberg** [comment]
Some further info in case it's useful.

```
$ strace python3 classify_image.py 2>&1 | grep openat | grep -i cublas | grep opt

openat(AT_FDCWD, "/opt/nvidia/cuda/lib64/libcublas.so.10", O_RDONLY|O_CLOEXEC) = 4
openat(AT_FDCWD, "/opt/nvidia/cuda/lib64/libcublasLt.so.10", O_RDONLY|O_CLOEXEC) = 4
```

It is looking in /opt as it should - great. But then I don't understand why it's not using libcublas from there. In that directory there are `libcublas.so`, `libcublas.so.10`, `libcublas.so.10.1`, `libcublas.so.10.1.0.105`, `libcublasLt.so`,  `libcublasLt.so.10`, `libcublasLt.so.10.1.0.105`

Then I check if it is looking in /usr:
```
$ strace python3 classify_image.py 2>&1 | grep openat | grep -i cublas | grep '/usr/lib64/libcublas'

openat(AT_FDCWD, "/usr/lib64/libcublas.so.10.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

It is looking there. So for now I worked around the issue as follows:
```
ln -s /opt/nvidia/cuda/lib64/libcublas.so.10.1 /usr/lib64/libcublas.so.10.1
```

**Comment by mengdong** [comment]
I cherry-picked the 2 commits into Tensorflow master, however, run into following errors:
`ERROR: /home/dongm/workspace/GITHUB/tensorflow/third_party/gpus/cuda_configure.bzl:131:1: Variable _GCC_HOST_COMPILER_PATH is read only
Need help? See https://bazel.build/versions/master/docs/skylark/errors/read-only-variable.html
ERROR: error loading package '': in /home/dongm/workspace/GITHUB/tensorflow/tensorflow/workspace.bzl: Extension 'third_party/gpus/cuda_configure.bzl' has errors
ERROR: error loading package '': in /home/dongm/workspace/GITHUB/tensorflow/tensorflow/workspace.bzl: Extension 'third_party/gpus/cuda_configure.bzl' has errors
INFO: Elapsed time: 0.193s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)`

**Comment by kgizdov** [comment]
currently this seems to suffer the same deficiency as described in #28093

**Comment by chsigg** [comment]
Hopefully PR #28021 will fix the issue you are seeing.

**Comment by kyleabeauchamp** [comment]
Any updates here?

**Comment by nluehr** [comment]
Closing as 10.1 support has been implemented separately by @chsigg.

**Pull request closed without merging**